### PR TITLE
21 oauth webhook subscription

### PR DESCRIPTION
**Pull request by Tenki99**

### CODE REVIEW

File: webhook_server/app.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        user_queues[userid].put('File change detected')
+    return '', 200
+
+def stream_events(userid):
+    while True:
+        if userid in user_queues and not user_queues[userid].empty():
+            message = user_queues[userid].get()
+            yield f"data: {message}\n\n"
+        yield ":heartbeat\n\n"
+        time.sleep(20)
```
Comment (fqzz2000) [inline webhook_server/app.py:35]: Queueu.get() is blocking. No need time.sleep()

File: src/main.py [hunk 2]
```diff
@@ -3,51 +3,76 @@
...
-    args = parser.parse_args()
+def run_login_server():
+    login_app.run(debug=False, port=5000, use_reloader=False)
 
-    if args.command is None:
-        parser.print_help()
-    else:
-        args.func()
+def listen_for_events(url):
+    while True:
```
Comment (fqzz2000) [inline src/main.py:63]: Keep looping without waiting? How to terminate this thread?

File: src/login_server.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        print(oauth_result.access_token)
+        url = 'http://localhost:5001/'
+        data = {'token': oauth_result.access_token}
+        requests.post(url, json=data)
+        return 'Success'
+    except dropbox.oauth.DropboxOAuth2FlowError as e:
+        return 'Error: %s' % (e,)
+
+@login_app.route('/shutdown', methods=['POST'])
+def shutdown():
```
Comment (fqzz2000) [inline src/login_server.py:33]: Nobody calls this API.

File: src/login_server.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+from flask import Flask, request, redirect, session
+import dropbox
+import requests
+import os
+
+login_app = Flask(__name__)
+login_app.secret_key = os.urandom(24)
+
+APP_KEY = 'p379vmpas0tf58c'
```
Comment (fqzz2000) [inline src/login_server.py:9]: put these global variables into a config file

File: src/main.py [hunk 5]
```diff
@@ -3,51 +3,76 @@
...
+import requests
+from loguru import logger
+from login_server import login_app
+from multiprocessing import Process, Queue
+from flask import Flask, request, redirect, session
+import dropbox
+import logging
 
 APP_KEY = "p379vmpas0tf58c"
+SUBSCRIBE_URL = "https://vcm-39026.vm.duke.edu:5002/events"
```
Comment (fqzz2000) [inline src/main.py:25]: put it into a global file

File: webhook_server/app.py (hunk 1 above)
Comment (Tenki99) [reply webhook_server/app.py:35]: This sleep is mainly for reduce the frequency of sending heartbeat. Could adjust the time of sleep according to the performance in later testing (likely to reduce the time of sleep)

**Review by fqzz2000 — APPROVED** [review]
LGTM

**Review by fqzz2000 — APPROVED** [review]

**Pull request merged**